// @eds/vanilla/dropdown/Menu.js
export class Menu {
  constructor(element) {
    this.dom = {
      menu: element,
      items: element.querySelectorAll('.item'),
      emptyMessage: element.querySelector('.empty'),
      highlightItem: element.querySelector('.highlight'),
    };
    this.max = element.getAttribute('data-display-max') || 5;
  }

  init() {
    this.limitDisplayMaxItems();
  }

  unHighlightItem(item) {
    item.classList.remove('highlight');
  }

  toggle() {
    if (this.dom.menu.classList.contains('visible')) {
      this.hide();
    } else {
      this.show();
    }
  }

  show() {
    this.dom.menu.classList.add('visible');
  }

  hide() {
    this.dom.menu.classList.remove('visible');
  }

  limitDisplayMaxItems() {
    Array.from(this.dom.items).forEach(item => item.classList.add('hidden'));
    Array.from(this.dom.items).forEach((item, i) => {
      if (i < this.max) item.classList.remove('hidden');
    });
  }

  showList(list) {
    // hide all items from the menu list
    Array.from(this.dom.items).forEach(item => item.classList.add('hidden'));
    // remove the hidden class from those that appear in the filteredList
    Array.from(list).forEach((item, i) => {
      if (i < this.max) item.classList.remove('hidden');
    });
    if (list.length === 0) {
      this.showEmptyState();
    } else {
      this.hideEmptyState();
    }
  }

  showEmptyState() {
    this.dom.emptyMessage.classList.add('visible');
  }

  hideEmptyState() {
    this.dom.emptyMessage.classList.remove('visible');
  }

  unHighlightAll() {
    Array.from(this.dom.items).forEach(item => this.unHighlightItem(item));
  }

  getHighlightedItem() {
    return this.dom.menu.querySelector('.highlight');
  }
}
