// @eds/vanilla/pill/Pill.js
export class Pill {

  /**
   * Setup pill properties
   * @param {HTMLElement} element - The DOM element of the pill
   */
  constructor(element) {
    this.dom = {
      pill: element,
    };
  }

  /**
   * Initialize the pill
   */
  init() {
    this.addEventListeners();
  }

  /**
   * Destroy the pill and associated events
   */
  destroy() {
    this.removeEventListeners();
    if (this.dom.pill && this.dom.pill.parentNode) {
      this.dom.pill.parentNode.removeChild(this.dom.pill);
    }
  }

  /**
   * Add event listeners
   * @private
   */
  addEventListeners() {
    this.dom.pill.addEventListener('click', () => this.onClick(), false);
  }

  /**
   * Remove event listeners
   * @private
   */
  removeEventListeners() {
    this.dom.pill.removeEventListener('click', () => this.onClick(), false);
  }

  /**
   * What happens on the click event
   * @private
   */

  onClick() {
    this.toggle();
    this.remove();
  }

  /**
   * Toggle the pill
   */
  toggle() {
    if (this.dom.pill.classList.contains('unselected')) {
      this.select();
    } else {
      this.unselect();
    }
  }

  /**
   * Select the pill
   */
  select() {
    this.dom.pill.classList.remove('unselected');
  }

  /**
   * Unselect the pill
   */
  unselect() {
    this.dom.pill.classList.add('unselected');
  }

  /**
   * Remove the pill
   */
  remove() {
    if (this.dom.pill.classList.contains('removable')) {
      this.destroy();
    }
  }
}
