// @eds/vanilla/sign-in/SignIn.js
export class SignIn {
  constructor(element, loggedInUrl = '/') {
    this.dom = {
      signin: element,
      reveal: element.querySelector('.reveal'),
      inputPassword: element.querySelector('#password'),
      form: element.querySelector('form'),
      error: element.querySelector('.error'),
      welcome: element.querySelector('.welcome'),
      loggedInUrl,
    };
    this.listeners = {
      click: () => this.revealPassword(),
      blur: (e) => this.validateFieldOnBlur(e),
      submit: (e) => this.validateFormOnSubmit(e),
    };
    this.messages = {
      valueMissing: 'Please fill out this field.',
      typeMismatch: {
        email: 'Please use a valid email.',
        fallback: 'Please use the correct input type.'
      },
      patternMismatch: {
        signum: 'Please use a valid signum.',
        fallback: 'Please match the requested format.'
      },
      fallback: 'Please enter a correct value for this field.'
    };
    this.props = {
      validFields: []
    };
  }

  init() {
    this.removeNativeValidation();
    this.addEventListeners();
    // this.showGenericError();
  }

  destroy() {
    this.removeEventListeners();
  }

  addEventListeners() {
    this.dom.reveal.addEventListener('click', this.listeners.click, false);
    this.dom.signin.addEventListener('blur', this.listeners.blur, true);
    this.dom.signin.addEventListener('submit', this.listeners.submit, true);
  }

  removeEventListeners() {
    this.dom.reveal.removeEventListener('click', this.listeners.click, false);
    this.dom.signin.removeEventListener('blur', this.listeners.blur, false);
    this.dom.signin.removeEventListener('submit', this.listeners.submit, false);
  }

  revealPassword() {
    const icon = this.dom.signin.querySelector('.reveal .icon');
    const message = this.dom.reveal.querySelector('.message');
    if (this.dom.inputPassword.type === 'password') {
      this.dom.inputPassword.type = 'text';
      icon.classList.remove('icon-eye');
      icon.classList.add('icon-eye-solid');
      message.innerText = 'Hide password';
    } else {
      this.dom.inputPassword.type = 'password';
      icon.classList.remove('icon-eye-solid');
      icon.classList.add('icon-eye');
      message.innerText = 'Show password';
    }
  }

  removeNativeValidation() {
    this.dom.form.setAttribute('novalidate', true);
  }

  validateFieldOnBlur(e) {
    const inputField = e.target;
    const field = inputField.parentNode;
    const hint = field.querySelector('.hint');

    if (field.classList.contains('validate')) {
      const error = this.getErrorType(inputField);

      if (error && error !== this.messages.valueMissing) {
        hint.innerText = error;
        inputField.classList.add('invalid');
      } else {
        inputField.classList.remove('invalid');
        inputField.classList.add('hidden');
      }
    }
  }

  validateFormOnSubmit(e) {
    const formFields = e.target.querySelectorAll('.field');

    Array.from(formFields).forEach((field, i) => {
      const inputField = field.querySelector('input');
      const hint = field.querySelector('.hint');
      const error = this.getErrorType(inputField);

      if (error) {
        hint.innerText = error;
        inputField.classList.add('invalid');
        e.preventDefault();
        this.props.validFields[i] = false;
      } else {
        inputField.classList.remove('invalid');
        inputField.classList.add('hidden');
        this.props.validFields[i] = true;
      }
    });

    const allFieldsValid = this.props.validFields.every(state => state === true);

    // This is just for demo purposes
    if (allFieldsValid) {
      this.successSignIn();
    }
  }

  getErrorType(field) {
    if (field.disabled
      || field.type === 'file'
      || field.type === 'reset'
      || field.type === 'submit'
      || field.type === 'button') {
      return;
    }
    const validity = field.validity;
    if (validity.valid) return;
    if (validity.valueMissing) return this.messages.valueMissing;
    if (validity.typeMismatch) {
      if (field.type === 'email') return this.messages.typeMismatch.email;
      return this.messages.typeMismatch.fallback;
    }
    if (validity.patternMismatch) {
      if (field.classList.contains('signum')) return this.messages.patternMismatch.signum;
      return this.messages.patternMismatch.fallback;
    }
    return this.messages.fallback;
  }

  successSignIn() {
    this.hideSignInForm();
    this.showWelcomeScreen();
    // This is just for demo purposes
    setTimeout(() => {
      window.location.href = this.dom.loggedInUrl;
    }, 3000);
  }

  showWelcomeScreen() {
    this.dom.welcome.classList.remove('hidden');
  }

  hideSignInForm() {
    this.dom.form.classList.add('hidden');
  }
}
