// @eds/vanilla/date-field/Datepicker.js
import { Calendar } from './Calendar';

export class Datepicker {
  constructor(element) {
    this.dom = {
      datepicker: element,
      calendar: element.querySelector('.calendar'),
      button: element.querySelector('.btn, .clickable'),
      input: element.querySelector('input[type="text"]'),
      days: null,
      hiddenInput: element.querySelector('input[type="hidden"]'),
    };
    this.listeners = {
      ifClickedOutside: this.ifClickedOutside.bind(this),
      selectDay: this.selectDay.bind(this),
      toggleCalendar: this.toggleCalendar.bind(this),
      listenInput: this.listenInput.bind(this),
    };

    this.calendar = null;
    this.selectedDate = null;
  }

  init() {
    this.calendar = new Calendar(this.dom.calendar);
    this.calendar.init();
    this.dom.days = this.dom.calendar.querySelectorAll('td');
    this.addEventListeners();
  }

  destroy() {
    this.removeEventListeners();
  }

  removeEventListeners() {
    const _self = this;

    this.dom.button.removeEventListener('click', () => _self.listeners.toggleCalendar(), false);
    Array.from(this.dom.days).forEach(day => {
      day.removeEventListener('click', (e) => _self.listeners.selectDay(e), false);
    });
    this.dom.input.removeEventListener('click', () => _self.listeners.toggleCalendar(), false);
    this.dom.input.removeEventListener('keyup', (e) => _self.listeners.listenInput(e), false);
    document.removeEventListener('click', _self.listeners.ifClickedOutside, false);
  }

  addEventListeners() {
    const _self = this;

    this.dom.button.addEventListener('click', () => _self.listeners.toggleCalendar(), false);
    Array.from(this.dom.days).forEach(day => {
      day.addEventListener('click', (e) => _self.listeners.selectDay(e), false);
    });
    this.dom.input.addEventListener('click', () => _self.listeners.toggleCalendar(), false);
    this.dom.input.addEventListener('keyup', (e) => _self.listeners.listenInput(e), false);
    document.addEventListener('click', _self.listeners.ifClickedOutside, false);
  }

  isValidDate(dateString) {
    const regEx = /^\d{4}-\d{2}-\d{2}$/;
    // Invalid format
    if (!dateString.match(regEx)) {
      return false;
    }
    const d = new Date(dateString);
    // Invalid date (or this could be epoch)
    if (!(d | 0)) {
      return false;
    }
    return d.toISOString().slice(0, 10) === dateString;
  }

  listenInput() {
    const typedInput = this.dom.input.value;
    if (typedInput.length === 10) {
      if (this.isValidDate(typedInput)) {
        this.calendar.setDataCalendar(new Date(typedInput));
        const dateArray = typedInput.split('-');
        this.selectedDate = {
          year: dateArray[0],
          month: dateArray[1],
          day: dateArray[2],
        };
        this.setSelectedDateCSS();
      } else {
        this.invalidDateFormat();
      }
    }
  }

  // set date to today when detecting an invalid date format
  invalidDateFormat() {
    const today = new Date();
    const todayYYYYMMDD = this.calendar.yyyymmdd(today).split('-');
    this.dom.input.value = todayYYYYMMDD.join('-');
    this.selectedDate = {
      year: todayYYYYMMDD[0],
      month: todayYYYYMMDD[1],
      day: todayYYYYMMDD[2],
    };
    this.calendar.setDataCalendar(new Date(today));
    this.calendar.removeAllSelectedDays();
    this.setSelectedDateCSS();
  }

  setSelectedDateCSS() {
    this.calendar.removeAllSelectedDays();

    // convert to string for equality check
    const selectedDay = this.selectedDate ?
      this.selectedDate.day :
      (new Date(this.dom.input.value)).getDate() + '';

    Array.from(this.dom.days).some(day => {
      if (day.innerText === selectedDay) {
        day.classList.add('selected');
        return true;
      }
    });
  }

  selectDay(event) {
    // TODO: Is this just for testing or should this check be in production?
    if (!event.target.classList.contains('dummy')) {
      this.calendar.removeAllSelectedDays();
      this.selectedDate = {
        year: this.calendar.calendar.year,
        month: this.calendar.calendar.month,
        day: event.target.innerText,
      };
      const formattedDate = this.calendar.formatDate(this.selectedDate);
      this.dom.input.value = this.dom.hiddenInput.value = formattedDate;
      if (event.target.tagName === 'TD') {
        event.target.classList.add('selected');
      } else {
        event.target.parentNode.classList.add('selected');
      }
      this.dom.input.dispatchEvent(new Event('change'));
      this.hide();
    }
  }

  hide() {
    this.dom.calendar.classList.add('closed');
  }

  show() {
    this.dom.calendar.classList.remove('closed');
  }

  toggleCalendar() {
    if (this.dom.calendar.classList.contains('closed')) {
      this.show();
    } else {
      this.hide();
    }

    const inputValue = this.dom.input.value;
    // not empty input
    if (inputValue) {
      if (this.isValidDate(inputValue)) {
        this.calendar.setDataCalendar(new Date(inputValue));
        this.setSelectedDateCSS();
      } else {
        this.invalidDateFormat();
      }
    }
  }

  ifClickedOutside(event) {
    if (!this.dom.datepicker.contains(event.target)) {
      this.hide();
    }
  }

}
