// @eds/vanilla/common/scripts/Appnav.js
export class Appnav {
  constructor(element) {
    this.dom = {
      appnav: element,
    };
    this.listeners = {
      toggleNavigation: () => this.toggle(),
      showNavigation: () => this.show(),
      hideNavigation: () => this.hide(),
    };
  }

  init() {
    this.addEventListeners();
  }

  destroy() {
    this.removeEventListeners();
  }

  addEventListeners() {
    document.addEventListener(
      'toggleNavigation', this.listeners.toggleNavigation, false);

    document.addEventListener(
      'showNavigation', this.listeners.showNavigation, false);

    document.addEventListener(
      'hideNavigation', this.listeners.hideNavigation, false);
  }

  removeEventListeners() {
    document.removeEventListener(
      'toggleNavigation', this.listeners.toggleNavigation, false);

    document.removeEventListener(
      'showNavigation', this.listeners.showNavigation, false);

    document.removeEventListener(
      'hideNavigation', this.listeners.hideNavigation, false);
  }

  toggle() {
    if (this.dom.appnav.classList.contains('hidden')) {
      this.show();
    } else {
      this.hide();
    }
  }

  show() {
    this.dom.appnav.classList.remove('hidden');
  }

  hide() {
    this.dom.appnav.classList.add('hidden');
  }

}
