// @eds/vanilla/common/scripts/Layout.js
import { App } from './App';
import { Syspanel } from './Syspanel';
import { Sysbar } from './Sysbar';

export class Layout {
  constructor(element) {
    this.dom = {
      body: element,
      app: element.querySelector('.app'),
      syspanel: element.querySelector('.syspanel'),
      sysbar: element.querySelector('.sysbar')
    };
    this.app = new App(this.dom.app);
    this.syspanel = new Syspanel(this.dom.syspanel);
    this.sysbar = new Sysbar(this.dom.sysbar);
    this.listeners = {
      resize: () => this.contentBehaviour(),
    };
    this.events = {
      showNavigation: new CustomEvent('showNavigation'),
      hideNavigation: new CustomEvent('hideNavigation'),
      toggleNavigation: new CustomEvent('toggleNavigation'),
    };
  }

  init() {
    this.app.init();
    this.syspanel.init();
    this.sysbar.init();
    this.addEventListeners();
    this.contentBehaviour();
  }

  destroy() {
    this.removeEventListeners();
    this.app.destroy();
    this.syspanel.destroy();
    this.sysbar.destroy();
  }

  addEventListeners() {
    window.addEventListener('resize', this.listeners.resize, false);
  }

  removeEventListeners() {
    window.removeEventListener('resize', this.listeners.resize, false);
  }

  contentBehaviour() {
    const viewport = {
      sm: 480,
      md: 768,
      lg: 1024,
    };
    const w = window.outerWidth;
    if (w <= viewport.sm) {
      document.dispatchEvent(this.events.hideNavigation);
    } else if (w <= viewport.md) {
      document.dispatchEvent(this.events.hideNavigation);
    } else if (w <= viewport.lg) {
      document.dispatchEvent(this.events.showNavigation);
    } else {
      document.dispatchEvent(this.events.showNavigation);
    }
  }

}
