// @eds/vanilla/multi-panel-tile/MultiPanelTile.js
export class MultiPanelTile {

  constructor(element) {
    this.dom = {
      multiPanelTile: element,
      leftPanel: element.querySelector('.left-panel'),
      midPanel: element.querySelector('.mid-panel'),
      rightPanel: element.querySelector('.right-panel'),
      leftPanelTriggers: element.querySelectorAll('.left-panel-trigger'),
      rightPanelTriggers: element.querySelectorAll('.right-panel-trigger'),
    };
    this.listeners = {
      toggleLeftPanel: () => {
        this.togglePanel(this.dom.leftPanel);
        this.updatePanelBehavior();
      },
      toggleRightPanel: () => {
        this.togglePanel(this.dom.rightPanel);
        this.updatePanelBehavior();
      },
    };
    this.flags = {
      panelLeft: false,
      panelRight: false,
    };
  }

  init() {
    this.addEventListeners();
    this.updatePanelBehavior();
  }

  destroy() {
    this.removeEventListeners();
  }

  addEventListeners() {
    Array.from(this.dom.leftPanelTriggers).forEach(trigger => {
      trigger.addEventListener('click', this.listeners.toggleLeftPanel, false);
    });
    Array.from(this.dom.rightPanelTriggers).forEach(trigger => {
      trigger.addEventListener('click', this.listeners.toggleRightPanel, false);
    });

    window.addEventListener('resize', () => this.updatePanelBehavior(), false);
  }

  removeEventListeners() {
    Array.from(this.dom.leftPanelTriggers).forEach(trigger => {
      trigger.removeEventListener('click', this.listeners.toggleLeftPanel, false);
    });
    Array.from(this.dom.rightPanelTriggers).forEach(trigger => {
      trigger.removeEventListener('click', this.listeners.toggleRightPanel, false);
    });

    window.removeEventListener('resize', () => this.updatePanelBehavior(), false);
  }

  togglePanel(panel) {
    if (!this.isPanelClosed(panel)) {
      this.hidePanel(panel);
    } else {
      this.showPanel(panel);
    }
  }

  isPanelClosed(panel) {
    return panel.classList.contains('hidden');
  }

  hidePanel(panel) {
    panel.classList.add('hidden');
  }

  showPanel(panel) {
    if (panel.classList.contains('left-panel')) {
      this.flags.panelLeft = true;
    }
    if (panel.classList.contains('right-panel')) {
      this.flags.panelRight = true;
    }
    panel.classList.remove('hidden');
  }

  updatePanelBehavior() {
    const leftPanel = this.dom.leftPanel;
    const midPanel = this.dom.midPanel;
    const rightPanel = this.dom.rightPanel;
    const pageWidth = document.body.offsetWidth;

    // mobile, tablet
    if (pageWidth < 769) {
      if (!this.isPanelClosed(leftPanel) && !this.isPanelClosed(rightPanel)) {
        this.hidePanel(leftPanel);
        this.hidePanel(midPanel);
      }

      if (this.isPanelClosed(leftPanel) && this.isPanelClosed(rightPanel)) {
        this.showPanel(midPanel);
      }

      if (!this.isPanelClosed(leftPanel) && this.isPanelClosed(rightPanel)) {
        this.hidePanel(midPanel);
        this.showPanel(leftPanel);
      }

      if (this.isPanelClosed(leftPanel) && !this.isPanelClosed(rightPanel)) {
        this.hidePanel(midPanel);
        this.showPanel(rightPanel);
      }

      // desktop, larger screen
    } else if (pageWidth < 1201) {
      this.showPanel(midPanel);
      if (!this.isPanelClosed(leftPanel) && !this.isPanelClosed(rightPanel)) {

        if (this.flags.panelLeft) {
          this.hidePanel(rightPanel);
        } else {
          this.hidePanel(leftPanel);
        }

        if (this.flags.panelRight) {
          this.hidePanel(leftPanel);
          this.showPanel(rightPanel);
        }
      }
    } else {
      this.showPanel(midPanel);
    }

    this.flags.panelLeft = false;
    this.flags.panelRight = false;
  }
}
